### PR TITLE
Implement PyPi package distribution metadata validation #1179

### LINKED ISSUE #1179: Implement PyPi package distribution metadata
# Task Description

We need to implement the PyPi metadata as specified [here](https://packaging.python.org/specifications/core-metadata/). The implementation of this metadata will have to check the validity of the data, as described in the specification.

The implementation should be based on the `pypi-layout-provider` branch.

# Task Relationships

This is a sub-task of #807.

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @fuss86

### PR DESCRIPTION
**Pull request by sainalshah**
# Pull Request Description

This pull request closes #1179

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [x] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [x] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

### CODE REVIEW

**Comment by carlspring** [comment]
Hi @sainalshah !

Thank you very much for your contribution!

Would you mind signing the ICLA, as described in the [Contributing](https://github.com/strongbox/strongbox/blob/master/CONTRIBUTING.md) page?

Also, please, feel free to join our [chat channel](https://chat.carlspring.org/), if you'd like to learn more about the project and/or like to find out what else you could help with.

Kind regards,

Martin

**Comment by carlspring** [comment]
@Firstmol ,

Would you like to help review this? :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/domain/PypiPackageInfo.java [hunk 1]
```diff
@@ -98,4 +114,35 @@ public String getPlatform()
...
+    public enum SupportedMetadataVersionEnum
+    {
+        VERSION_1_0("1.0"),
+        VERSION_1_1("1.1"),
+        VERSION_1_2("1.2"),
+        VERSION_2_1("2.1");
+
+        private String version;
+
+        public String getVersionString(){
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/domain/PypiPackageInfo.java]: Cosmetic: Would you mind following our [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html)? Braces are on new lines. Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/domain/PypiPackageInfo.java [hunk 2]
```diff
@@ -98,4 +114,35 @@ public String getPlatform()
...
+            {
+                if (metadataVersionEnum.getVersionString().equals(version))
+                {
+                    return metadataVersionEnum;
+                }
+            }
+            throw new IllegalArgumentException("Unsupported Metadata version: " + version);
+        }
+
+        SupportedMetadataVersionEnum(String version){
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/domain/PypiPackageInfo.java]: Cosmetic: Would you mind following our [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html)? Braces are on new lines. Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/util/PypiMetadataParser.java [hunk 3]
```diff
@@ -39,7 +45,11 @@ private PypiPackageInfo populateAnnotatedFields(PypiPackageInfo object, Map<Stri
                 if (field.isAnnotationPresent(PypiMetadataKey.class))
                 {
                     PypiMetadataKey pypiMetadataKey = field.getAnnotation(PypiMetadataKey.class);
-                    field.set(object, keyValueMap.get(pypiMetadataKey.name()));
+                    if (pypiMetadataKey.name().equals("Metadata-Version")) {
+                        field.set(object, PypiPackageInfo.SupportedMetadataVersionEnum.getVersionEnum(keyValueMap.get(pypiMetadataKey.name())));
+                    } else {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/util/PypiMetadataParser.java]: Cosmetic: Braces are new lines.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/util/PypiMetadataParser.java [hunk 4]
```diff
@@ -39,7 +45,11 @@ private PypiPackageInfo populateAnnotatedFields(PypiPackageInfo object, Map<Stri
                 if (field.isAnnotationPresent(PypiMetadataKey.class))
                 {
                     PypiMetadataKey pypiMetadataKey = field.getAnnotation(PypiMetadataKey.class);
-                    field.set(object, keyValueMap.get(pypiMetadataKey.name()));
+                    if (pypiMetadataKey.name().equals("Metadata-Version")) {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/util/PypiMetadataParser.java]: Cosmetic: Braces are new lines.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java [hunk 5]
```diff
@@ -5,12 +5,19 @@
...
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.test.context.ContextConfiguration;
 
+import javax.validation.ConstraintViolationException;
 import java.io.FileInputStream;
 import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
 
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (mslowiak) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java]: We decided to switch from junit and hamcrest to AssertJ.
Please change assertions to AssertJ way.
See: https://github.com/strongbox/strongbox/issues/1456

**Comment by sainalshah** [comment]
@carlspring @mslowiak Will do the requested changes and update soon. Bit busy at the moment due to work commitments.

**Comment by carlspring** [comment]
@sainalshah ,

Take your time! We're really interested in your help with this and would like to invite you to 
our [chat channel](https://chat.carlspring.org/).

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/util/PypiPackageInfoValidator.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package org.carlspring.strongbox.util;
+
+import org.carlspring.strongbox.domain.PypiPackageInfo;
+import org.springframework.stereotype.Component;
+
+import javax.validation.*;
+import java.util.Set;
+
+@Component
+public class PypiPackageInfoValidator {
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/main/java/org/carlspring/strongbox/util/PypiPackageInfoValidator.java]: Cosmetic: Could you please, re-indent according to our [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html)? Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java [hunk 7]
```diff
@@ -42,22 +54,123 @@ public void testParseFileToPypiMetadataDto()
...
-        assertThat(testDao.getLicense()).isEmpty();
-        assertThat(testDao.getDescriptionContentType()).isEmpty();
-        assertThat(testDao.getDescription()).isEmpty();
-        assertThat(testDao.getPlatform()).isEmpty();
+        assertThrows(ConstraintViolationException.class, () -> {
+            pypiMetadataParser.parseMetadataFile(new FileInputStream("src/test/resources/org.carlspring.strongbox.util/PKG-INFO-empty-values"));
+        });
+    }
+
+    @Test
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java]: Cosmetic: Could you please, re-indent your changes here according to our [coding convention](https://strongbox.github.io/developer-guide/coding-convention.html)? Thanks! :)

Comment (sainalshah) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java]: @carlspring code is reformatted to follow coding convention. Please check again. Thanks!

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java (hunk 5 above)
Comment (sainalshah) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java]: @mslowiak Test cases are updated to use AssertJ fluent assertions. Thanks!

Comment (mslowiak) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-pypi-layout-provider/src/test/java/org/carlspring/strongbox/util/PypiMetadataParserTest.java]: @sainalshah thank you for your follow up, good job :)

**Comment by sainalshah** [comment]
> Hi @sainalshah !
> 
> Thank you very much for your contribution!
> 
> Would you mind signing the ICLA, as described in the [Contributing](https://github.com/strongbox/strongbox/blob/master/CONTRIBUTING.md) page?
> 
> Also, please, feel free to join our [chat channel](https://chat.carlspring.org/), if you'd like to learn more about the project and/or like to find out what else you could help with.
> 
> Kind regards,
> 
> Martin

@carlspring I've sent the ICLA document.
I'm available on chat channel with username msainal.

Thank you!

**Comment by carlspring** [comment]
Thank you! I have received it!

Let.me know when you think it's ready for merging!

**Comment by sainalshah** [comment]
I think I'm done for this PR.

Probably we have to create a sub task to implement Version Normalization as mentioned [here](https://www.python.org/dev/peps/pep-0440/#normalization)

**Comment by carlspring** [comment]
Great, thanks!

Would you like to create such a task in our issue tracker along with a proper description?

We have a number of other tasks in the PyPi master task as well. It would be most really nice, if you would like to work on them as well, as follow-up tasks.

**Comment by carlspring** [comment]
Thank you for your contribution! :)

**Comment by sainalshah** [comment]
> Great, thanks!
> 
> Would you like to create such a task in our issue tracker along with a proper description?

Sure thing! I'll follow up by creating the task with necessary details.

> We have a number of other tasks in the PyPi master task as well. It would be most really nice, if you would like to work on them as well, as follow-up tasks.

Sure, I'm happy to help out on other tasks as well.

Thanks for accepting the PR :)

**Pull request merged**